libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,476 @@
+            const auto distance = GetQuantizationDistance(
+               peakIndices, peakValues, odf.size(), numTatums, lag);
+            return std::make_pair(
+               numTatums, OnsetQuantization { distance, lag, numTatums });
+         });
+      return quantizations;
+   }();
+
+   const auto bestFitIt = std::min_element(
+      quantizations.begin(), quantizations.end(),
+      [](const std::pair<int, OnsetQuantization>& a,
+         const std::pair<int, OnsetQuantization>& b) {
+         return a.second.error < b.second.error;
+      });
+
+   // Read the normalized distance, a value between 0 and 1, as a probability.
+   const auto error = bestFitIt->second.error;
+   const auto contrast = GetExperimentContrastMeasure(quantizations, error);
+   // The higher the contrast, the more likely the recording is a loop.
+   const auto score = 1 - error * (1 - contrast);
+
+   const auto mostLikelyNumTatums = bestFitIt->first;
+   const auto lag = bestFitIt->second.lag;
+
+   return { error, lag, mostLikelyNumTatums, score };
+}
+
+std::optional<TimeSignature>
+GetTimeSignature(const BarDivision& barDivision, int numTatums)
+{
+   const auto numBeats = barDivision.numBars * barDivision.beatsPerBar;
+   const auto tatumsPerBeat = 1. * numTatums / numBeats;
+   switch (barDivision.beatsPerBar)
+   {
+   case 2:
+      return (tatumsPerBeat == 3) ? TimeSignature::SixEight :
+                                    TimeSignature::TwoTwo;
+   case 3:
+      return TimeSignature::ThreeFour;
+   case 4:
+      return TimeSignature::FourFour;
+   default:
+      // Since the bar-division hypotheses are based on these very time
+      // signatures, this should never happen. But the code would need to be
+      // structured more clearly to really have confidence and risk crashing.
+      // For now we accept that the time-signature cannot be recognized in
+      // release builds.
+      assert(false);
+      return std::nullopt;
+   }
+}
+
+double
+GetTimeSignatureLikelihood(const std::optional<TimeSignature>& ts, double bpm)
+{
+   // TODO these were taken from a rather old PoC - review.
+   if (!ts.has_value())
+      return 0.;
+
+   static const std::unordered_map<TimeSignature, double> expectedBpms {
+      { TimeSignature::TwoTwo, 115. },
+      { TimeSignature::FourFour, 115. },
+      { TimeSignature::ThreeFour, 140. },
+      { TimeSignature::SixEight, 64. },
+   };
+
+   static const std::unordered_map<TimeSignature, double> bpmStdDevs {
+      { TimeSignature::TwoTwo, 25. },
+      { TimeSignature::FourFour, 25. },
+      { TimeSignature::ThreeFour, 25. },
+      { TimeSignature::SixEight, 15. },
+   };
+
+   // Add a slight bias towards 4/4, which is the most common time signature.
+   static const std::unordered_map<TimeSignature, double> tsPrior {
+      { TimeSignature::TwoTwo, .1 },
+      { TimeSignature::FourFour, .45 },
+      { TimeSignature::ThreeFour, .2 },
+      { TimeSignature::SixEight, .25 },
```

Comment:
What is the source of all the numbers in these three maps?

Reply:
The data in the third map -- however derived -- might be used in that suggested Bayesian formula.

Reply:
You're right, I am using probabilistic measures that may be more clearly be brought together with a clear formulation of the Bayes theorem. I just created a ticket for that: https://github.com/audacity/audacity/issues/5826

Reply:
I still don't know the source of the numbers.